### ISSUE TITLE
Support for inline labels for plots

### ISSUE THREAD

**Issue by RemDelaporteMathurin** [matplotlib/matplotlib#19406]
### Problem

I would like to be able to use inline labels for plots instead of having the legend in a box.

### Proposed Solution
A workaround was found [here](https://stackoverflow.com/questions/16992038/inline-labels-in-matplotlib) but it's still not ideal.

![image](https://user-images.githubusercontent.com/40028739/106366558-f4222c80-633c-11eb-8ed7-bea17d301180.png)

Contour plots can handle inline labels by removing parts of the contour where the label is. Could the same thing be available for "normal" plots ?

**Comment by jklymak** [comment]
There is a whole package based on that stack overflow post: https://github.com/cphyc/matplotlib-label-lines  We could invite them to contribute to mainline matplotlib if they think it stable enough.  Looks like it has been around for a while, so maybe they have a stable API for it.

**Comment by RemDelaporteMathurin** [comment]
@jklymak the main drawback with this package is that it creates a white rectangle instead of removing the underlying path/line.
This is utterly ugly in some cases, whereas applying the same method as in contour.py would be grand

**Comment by RemDelaporteMathurin** [comment]
Also, it sometimes doesn't take rotations into account when applying custom x locations to the labels.
![image](https://user-images.githubusercontent.com/40028739/106367356-56316080-6342-11eb-9434-beb0c5d81278.png)

**Comment by jklymak** [comment]
I see, well, that sounds like an ambitious project, but if there is a will there is a way.  I would still encourage someone to write it as a standalone new artist first if possible to nail down the API.

**Comment by RemDelaporteMathurin** [comment]
Could you explain what you mean by "a standalone new artist" ?

**Comment by jklymak** [comment]
I mean I will be (pleasantly) surprised if you could rip parts out of a Line2D artist as easily as contour.   But I haven't looked into carefully.  If that is true, you would probably want to create a new plotting routine to take the place of "plot"

**Comment by RemDelaporteMathurin** [comment]
Tbh I am far from being an expert of what happens behind the scenes of matplotlib.

With contours, clabel is called after contour without really replacing it. So maybe a similar thing could be done a posteriori to "plot".

But then again, I have no idea whatsoever how we could break a 2D line. Could replacing the x, y values by NaN where the label lies do the trick ? Perhaps not the most elegant solution ....

**Comment by jklymak** [comment]
> But then again, I have no idea whatsoever how we could break a 2D line. Could replacing the x, y values by NaN where the label lies do the trick ? Perhaps not the most elegant solution ....

Well, you need to put breaks in lines if there is a long line between two points, so that adds considerable complexity because you not only have to remove points, you need to add them as well.

**Comment by RemDelaporteMathurin** [comment]
I understand. I kinda consider the following strategy for a `clabel` "equivalent" for `plot`:
1) remove the points where the label must stand (based on some pixel width)
2) if points around the label are too far apart (to be defined), perform some kind of linear interpolation to add some points "around" the label

I think it should be some kind of adaptation of `add_label_near()` to `plot`
https://github.com/matplotlib/matplotlib/blob/cd0c551ecdd3fa6939a7be882c5de3bae00a17e6/lib/matplotlib/contour.py#L454-L522

**Comment by RemDelaporteMathurin** [comment]
@jklymak the following prototype seems to produce interesting results while being relatively simple.
It obviously has some caveats but I think this could be the way to go. What do you think ?

```python
import matplotlib.pyplot as plt
import numpy as np
from scipy.interpolate import interp1d


def inline_label(line, x, s, width=10):
    # get line data
    x_data = line.get_xdata()
    y_data = line.get_ydata()
    interpolated_data = interp1d(x_data, y_data)

    x_left = x - width/2
    x_right = x + width/2
    x_closest_left = x_data[(x_left > x_data)][-1]
    x_closest_right = x_data[(x_right < x_data)][0]

    # add points not enough points at location
    nb_points_at_location = len(x_data[(x_data > x_left) & (x_data < x_right)])
    threshold = 10  # to be determined in a more clever way
    if nb_points_at_location < threshold:
        index = np.where(x_data == x_closest_left)[0] + 1
        inserted_x = np.linspace(x_left, x_right, 30)
        inserted_y = interpolated_data(np.linspace(x_left, x_right, 30))
        x_data = np.insert(x_data, index, inserted_x)
        y_data = np.insert(y_data, index, inserted_y)

    # replace points by nan
    indexes = (x_left <= x_data) & (x_data <= x_right)
    x_data[indexes] = np.nan
    y_data[indexes] = np.nan

    # insert additional points arround label
    index_left = np.where(x_data == x_closest_left)[0]
    x_data = np.insert(x_data, index_left, x_left - width/2)
    y_data = np.insert(y_data, index_left, interpolated_data(x_left - width/2))

    index_right = np.where(x_data == x_closest_right)[0]
    x_data = np.insert(x_data, index_right, x_right + width/2)
    y_data = np.insert(y_data, index_right, interpolated_data(x_right + width/2))

    # modify xdata and ydata of 2D line
    line.set_xdata(x_data)
    line.set_ydata(y_data)

    # add label
    y_label = interpolated_data(x_left)
    plt.text(x_left, y_label, s)
    return


data_y = np.linspace(1, 3, num=50)
data_x = np.linspace(1, 50, num=50)
line, = plt.plot(data_x, data_y)
inline_label(line, 25, s="label_1", width=5)

data_y = np.linspace(2, 5, num=8)
data_x = np.linspace(1, 50, num=8)
line, = plt.plot(data_x, data_y)
inline_label(line, 25, s="label_2", width=5)

data_y = np.linspace(5, 7, num=5)
data_x = np.linspace(1, 50, num=5)
line, = plt.plot(data_x, data_y)
inline_label(line, 25, s="label_3", width=5)
plt.show()
```
![image](https://user-images.githubusercontent.com/40028739/106382126-d0eb9180-63bd-11eb-9ba7-db9a28c96049.png)

**Comment by cphyc** [comment]
Hi,
I'm the maintainer of https://github.com/cphyc/matplotlib-label-lines (mentioned above). This is just to add that I would be fine with merging the package into matplotlib. Quite a lot of work has been done to ensure it works in as many situations as possible, e.g. by handling different axis types (numeric, dates, etc.). It is not perfect, but it probably is a good start.

As mentioned by @RemDelaporteMathurin, it works by adding a white patch where the label is supposed to be located. This is quite ugly and should be replaced with his solution. It should be possible to reuse some of matplotlib's code (I am thinking about the `plt.text` API to add a `FancyBboxPatch`.

Anyway, I would be OK if the code from my package was included in matplotlib or accept PRs.

**Comment by RemDelaporteMathurin** [comment]
@cphyc perhaps the word "ugly" was a bit strong. It is really fine in most cases for sure and I have been using it for a while!
But when it comes to overlapping lines, or axes grid, then the beauty of it all starts to fade away I'm afraid.. :-( 

So glad to hear you're interested in this !

**Comment by cphyc** [comment]
Yes, that's why I added the mention to `FancyBboxPatch`. So that users may draw a box around their label to hide extra lines, axes grid, etc. I don't believe there is any simple solution to this problem.
Maybe an alternative and possibly better solution would be to have a way of creating a contour around the label (as presented in the attached figure) instead of a box. Is that something that's doable?

![example](https://user-images.githubusercontent.com/5411875/106441441-56cd1280-647a-11eb-8805-644493c056d2.png)

**Comment by jklymak** [comment]
My personal preference, however, would be that this development takes place in label-lines, at least to start.  Its seems like a self-contained project.  If it gets to the point where you need to hack into matplotlib internals to get it work, or you think the API is nailed down, then, we could discuss adding to the main library.

**Comment by timhoffm** [comment]
@cphyc You can do contouring with `TextPath`/`PathPatch`. See the relevant code for "textpath supports mathtext & TeX" in https://matplotlib.org/devdocs/gallery/text_labels_and_annotations/demo_text_path.html.

However, having the line peek through at the whitespace between "Some" and "text" will probably look awkward. Also the line ends would get funny shapes. I think a bounding box / really cut out lines are the better solution.

**Comment by RemDelaporteMathurin** [comment]
@timhoffm I agree with you. Cut the lines is the **ultimate solution** for sure. But the solution proposed by @cphyc was a simple enough to be implemented easily and works in most cases.

Do you think cutting out the lines would be easy to implement ?
